crates/objc2/src/runtime/nsobject.rs

```diff
@@ -155,6 +163,36 @@ pub unsafe trait NSObjectProtocol {
     // For example, something may have a return type of `NSString`, while
     // behind the scenes they really return `NSMutableString` and expect it to
     // not be modified.
+
+    /// Checks if the object is a member of the class.
+    ///
+    /// See [Apple's documentation][apple-doc] for more details on what you
+    /// may (and what you may not) do with this information.
+    ///
+    /// [apple-doc]: https://developer.apple.com/documentation/objectivec/1418956-nsobject/1418766-ismemberofclass
+    #[doc(alias = "isMemberOfClass")]
+    fn is_member_of<T: ClassType>(&self) -> bool
+    where
+        Self: Sized + Message,
+    {
+        self.__isMemberOfClass(T::class())
+    }
+
+    /// Tries to cast the object to the class `T` by checking whether the object
+    /// is a direct member of the class (through [`is_member_of`](`Self::is_member_of`)).
+    fn member_cast<T: ClassType>(&self) -> Option<&T>
+    where
+        Self: Sized + Message + 'static,
+        T: 'static,
+    {
+        if self.is_member_of::<T>() {
+            // SAFETY: self is a direct member of the specified (`T`) class.
+            // Therefore casting should be safe.
+            Some(unsafe { &*(self as *const Self).cast::<T>() })
+        } else {
+            None
+        }
+    }
```

Comment:
Hmm, this one is _tricky_, and the reason I haven't added it yet is because I'm not 100% sure it would be sound! It seems like it _should_ be, since `&self` and `&T` are both immutable references, which avoids the main problem with `Id::member_cast`.

But in any case, it seems like it could just as safely use `__isKindOfClass` as `__isMemberOfClass`?

Reply:
> But in any case, it seems like it could just as safely use __isKindOfClass as __isMemberOfClass?

I initially thought that having this restriction would reduce the common footgun described [here](https://developer.apple.com/documentation/objectivec/1418956-nsobject/1418511-iskindofclass#discussion). However, I also realize that our situation is slightly different since we have the concept of mutability here.

Reply:
No, the example they have:
```objective-c
// DO NOT DO THIS!
if ([myArray isKindOfClass:[NSMutableArray class]])
{
    // Modify the object
}
```

Is very, very similar to the problem that we have, and `isMemberOfClass:` suffers from it as well, e.g. the following is just as illegal (though likely just even worse, since `NSMutableArray` is in most runtimes a class cluster).
```objective-c
// DO NOT DO THIS!
if ([myArray isMemberOfClass:[NSMutableArray class]])
{
    // Modify the object
}
```

Reply:
In general I would really like to steer users away from `isMemberOfClass:`, since it depends on what is arguably implementation details (the exact class the type ends up as), and towards `isKindOfClass:`, which is more robust to that.

Reply:
> No, the example they have:

Well, I think we agree on that part? That's why I said "I initially thought". The problem is still there, but the proposed api should prevent "modification" through immutability no?